keps/0019-package-api-versioning.md

```diff
@@ -2,64 +2,370 @@
+2. Make the concept of the _operator version_ be the composition of the _app version_ and an _operator revision_. In this section, _package version_ and _operator version_ are the same thing (and will henceforth be called _version_), like in the current state of things with the singular version, and are also deterministic.
+
+   For example, in `2.3.0-1.0.0`:
+   - The _operator version_ (and by extension the _package version_, or just _version_) is `2.3.0-1.0.0`
+   - The _app version_ is `2.3.0`
+   - The _operator revision_ is `1.0.0`
+3. The _operator version_ wouldn't necessarily need to be explicitly set in operator or package metadata (`operator.yaml` and `index.yaml` respectively) given that the metadata contains fields for the _app version_ and the _operator revision_.
+
+It's important to note that **operator revision for different app versions are unrelated**. E.g., assuming two versions `2.3.0-1.0.0` and `3.0.0-1.0.0`, even though both have the same operator revision (`1.0.0`) they wouldn't necessarily have any commonality with regards to exclusively operator-related features and bug fixes. The operator revision progression is only meaningful within an app version's `major.minor` family, i.e. `2.3.x` or `3.0.x`.
+
+###### CLI UX
+
+| Concept                            | Flag          |
+| ------------------------           | ---------     |
+| Operator/package version (version) | --version     |
+| Application version                | --app-version |
+| Operator revision                  | --revision    |
+
+Assuming the following packages exist in the package registry:
+
+| Operator | Version     |
+| -------- | ----------- |
+| kafka    | 2.3.0-1.0.0 |
+| kafka    | 2.3.0-1.1.0 |
+| kafka    | 2.3.0-1.1.1 |
+| kafka    | 2.3.1-1.2.0 |
+| kafka    | 3.0.0-1.0.0 |
+| kafka    | 3.0.0-1.1.0 |
+| kafka    | 3.0.1-1.1.0 |
+| kafka    | 3.1.1-1.0.0 |
+
+**Install latest version of the Kafka operator**
+
+```
+kudo install kafka
+```
+
+Installs `kafka-3.1.1-1.0.0`.
+
+**Install the latest Kafka operator with a specific Kafka version**
+
+```
+kudo install kafka --app-version 2.3.0
+```
+
+Installs `kafka-2.3.0-1.1.1`.
+
+```
+kudo install kafka --app-version 3.0.0
+```
+
+Installs `kafka-3.0.0-1.1.0`.
+
+**Install a Kafka operator with a specific version**
+
+Let's say a user wants to use Apache Kafka `3.0.0` and even though there's a `3.0.0` with a `1.1.0` revision released, they want the operator with revision `1.0.0` due to reasons like:
+- revision `1.1.0` introduced a bug for which there's still no released fixes
+- revision `1.1.0` changed in a way that will require them to invest time and resources to adapt
+
+```
+kudo install kafka --version 3.0.0-1.0.0
+```
+
+**Install a Kafka operator providing a partial version**
```

Comment:
I very much like this part! 💯 
regardless of how we define it later, I think we should keep it